File: src/lib/alarm.ts
import { Shield, ShieldAlert, ShieldCheck, type LucideIcon } from "lucide-react";
import type { Channel } from "../components/PanelCard";

/** The three arm services we expose (HA `alarm_control_panel.*`). */
export type ArmService = "alarm_disarm" | "alarm_arm_home" | "alarm_arm_away";

const LABEL: Record<string, string> = {
  disarmed: "Disarmed",
  armed_home: "Armed — Home",
  armed_away: "Armed — Away",
  armed_night: "Armed — Night",
  armed_vacation: "Armed — Vacation",
  arming: "Arming…",
  pending: "Pending…",
  disarming: "Disarming…",
  triggered: "Triggered",
};

const SHORT: Record<string, string> = {
  disarmed: "Disarmed",
  armed_home: "Home",
  armed_away: "Away",
  armed_night: "Night",
  armed_vacation: "Vacation",
  triggered: "Triggered",
};

export interface AlarmView {
  /** Full human label, e.g. "Armed — Away". */
  label: string;
  /** Short label for the nav pill, e.g. "Away". */
  short: string;
  icon: LucideIcon;
  /** PULSE channel: green when settled, blue when armed, orange when triggered. */
  channel: Channel;
  armed: boolean;
  triggered: boolean;
  transitioning: boolean;
}

/**
 * Pure view-model for an alarm_control_panel state. Shared by `AlarmCard` and
 * the `SecurityStatusBar`/`TopNav` pill so the security read-out is identical
 * everywhere. State semantics mirror the original AlarmCard exactly.
 */
export function alarmView(state: string): AlarmView {
  const triggered = state === "triggered";
  const armed = state.startsWith("armed");
  const transitioning =
    state === "arming" || state === "pending" || state === "disarming";
  const icon = triggered ? ShieldAlert : armed ? Shield : ShieldCheck;
  const channel: Channel = triggered ? "streak" : armed ? "effort" : "recovery";
  return {
    label: LABEL[state] ?? state,
    short: SHORT[state] ?? LABEL[state] ?? state,
    icon,
    channel,
    armed,
    triggered,
    transitioning,
  };
}

/** The Off / Home / Away segmented control, in display order. */
export const ARM_BUTTONS: ReadonlyArray<{
  label: string;
  service: ArmService;
  /** The resulting state this button represents (for the `active` highlight). */
  state: string;
}> = [
  { label: "Off", service: "alarm_disarm", state: "disarmed" },
  { label: "Home", service: "alarm_arm_home", state: "armed_home" },
  { label: "Away", service: "alarm_arm_away", state: "armed_away" },
];

File: android/app/src/main/java/com/hawksnest/core/logic/Alarm.kt
package com.hawksnest.core.logic

/** Which shield glyph the UI should draw; kept pure (mapped to a Material icon in the UI). */
enum class AlarmGlyph { SHIELD, SHIELD_ALERT, SHIELD_CHECK }

private val LABEL = mapOf(
    "disarmed" to "Disarmed",
    "armed_home" to "Armed — Home",
    "armed_away" to "Armed — Away",
    "armed_night" to "Armed — Night",
    "armed_vacation" to "Armed — Vacation",
    "arming" to "Arming…",
    "pending" to "Pending…",
    "disarming" to "Disarming…",
    "triggered" to "Triggered",
)

private val SHORT = mapOf(
    "disarmed" to "Disarmed",
    "armed_home" to "Home",
    "armed_away" to "Away",
    "armed_night" to "Night",
    "armed_vacation" to "Vacation",
    "triggered" to "Triggered",
)

/**
 * Pure view-model for an `alarm_control_panel` state. Shared by the alarm card and the security
 * hero/status pill so the read-out is identical everywhere. Ported from `src/lib/alarm.ts`.
 */
data class AlarmView(
    /** Full human label, e.g. "Armed — Away". */
    val label: String,
    /** Short label for the nav pill, e.g. "Away". */
    val short: String,
    val glyph: AlarmGlyph,
    /** PULSE channel: green when settled, blue when armed, orange when triggered. */
    val channel: Channel,
    val armed: Boolean,
    val triggered: Boolean,
    val transitioning: Boolean,
)

fun alarmView(state: String): AlarmView {
    val triggered = state == "triggered"
    val armed = state.startsWith("armed")
    val transitioning = state == "arming" || state == "pending" || state == "disarming"
    val glyph = when {
        triggered -> AlarmGlyph.SHIELD_ALERT
        armed -> AlarmGlyph.SHIELD
        else -> AlarmGlyph.SHIELD_CHECK
    }
    val channel = when {
        triggered -> Channel.STREAK
        armed -> Channel.EFFORT
        else -> Channel.RECOVERY
    }
    return AlarmView(
        label = LABEL[state] ?: state,
        short = SHORT[state] ?: LABEL[state] ?: state,
        glyph = glyph,
        channel = channel,
        armed = armed,
        triggered = triggered,
        transitioning = transitioning,
    )
}

/** HA alarm-panel states where a command is still settling (exit delays, entry countdowns). */
val ALARM_TRANSITIONAL = setOf("arming", "disarming", "pending")

/** One segment of the Off / Home / Away control. [service] is the HA `alarm_control_panel.*` call. */
data class ArmButton(val label: String, val service: String, val state: String)

/** The Off / Home / Away segmented control, in display order. */
val ARM_BUTTONS: List<ArmButton> = listOf(
    ArmButton("Off", "alarm_disarm", "disarmed"),
    ArmButton("Home", "alarm_arm_home", "armed_home"),
    ArmButton("Away", "alarm_arm_away", "armed_away"),
)
